File: concepts/10_time_locale_and_runtime/01_durations_clocks_and_monotonic_time/python/test_02_units_and_clock_boundaries.py
"""时长单位和时钟边界。

共同问题：时长是否携带单位；墙上时间回拨时经过时间如何计算；
数值精度与单位转换由调用方还是类型系统约束。
"""

# polyglot-family: time_locale_and_runtime
# polyglot-concept: durations_clocks_and_monotonic_time
# polyglot-related: languages/python/stdlib/043-055_data_types/
# polyglot-related+: test_043_datetime_and_zoneinfo_complete_temporal_workflows.py
# polyglot-related: languages/python/stdlib/077-087_generic_operating_system_services/
# polyglot-related+: test_081_time_complete_epoch_formatting_clocks_and_timezones.py

import datetime

import pytest


def elapsed(start, finish):
    return finish - start


def test_timedelta_keeps_units_out_of_plain_numeric_arithmetic():
    duration = datetime.timedelta(seconds=1, milliseconds=250)

    assert duration.total_seconds() == 1.25
    assert duration / datetime.timedelta(milliseconds=250) == 5
    with pytest.raises(TypeError):
        duration + 1


def test_negative_timedelta_is_normalized_but_total_seconds_keeps_the_sign():
    duration = datetime.timedelta(microseconds=-1)

    assert duration.days == -1
    assert duration.seconds == 86_399
    assert duration.microseconds == 999_999
    assert duration.total_seconds() == -0.000001


def test_monotonic_samples_measure_progress_even_when_wall_clock_moves_back():
    wall_samples = (1_000.0, 995.0)
    monotonic_samples = (40.0, 42.5)

    assert elapsed(*wall_samples) == -5.0
    assert elapsed(*monotonic_samples) == 2.5

    # 这里注入读数而不修改系统时钟：回拨证明来自时钟语义，不依赖一次真实运行恰好发生校时。

File: concepts/10_time_locale_and_runtime/01_durations_clocks_and_monotonic_time/nodejs/test_02_units_and_clock_boundaries.mjs
// 时长单位和时钟边界。
// 共同问题：时长是否携带单位；墙上时间回拨时经过时间如何计算；
// 数值精度与单位转换由调用方还是类型系统约束。
//
// polyglot-family: time_locale_and_runtime
// polyglot-concept: durations_clocks_and_monotonic_time
// polyglot-related: languages/nodejs/node_core/10_diagnostics_and_testing/
// polyglot-related+: test_082_performance_timeline_observers_histograms_and_event_loop_metrics.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

function elapsed(start, finish) {
  return finish - start;
}

test('普通数值不携带时长单位，调用方必须显式换算', () => {
  const seconds = 1;
  const milliseconds = 250;

  assert.equal(seconds + milliseconds, 251);
  assert.equal(seconds * 1_000 + milliseconds, 1_250);
});

test('BigInt 纳秒保留整数精度但不能与 Number 隐式混算', () => {
  const duration = 1_000_000_001n;

  assert.equal(duration / 1_000_000n, 1_000n);
  assert.throws(() => duration + 1, TypeError);
});

test('单调读数在墙上时钟回拨样本中仍表达正向经过时间', () => {
  const wallSamples = [1_000, 995];
  const monotonicSamples = [40, 42.5];

  assert.equal(elapsed(...wallSamples), -5);
  assert.equal(elapsed(...monotonicSamples), 2.5);

  // 注入读数避免修改主机时钟；真实测量应从 performance.now 或 hrtime.bigint 取得单调样本。
});
